hlog/internal/mutil/body.go

```diff
@@ -0,0 +1,40 @@
+package mutil
+
+import (
+	"io"
+	"sync/atomic"
+)
+
+type byteCountReadCloser struct {
+	rc   io.ReadCloser
+	read int64
+}
+
+var _ io.ReadCloser = (*byteCountReadCloser)(nil)
+var _ io.WriterTo = (*byteCountReadCloser)(nil)
+
+func NewByteCountReadCloser(rc io.ReadCloser) *byteCountReadCloser {
+	return &byteCountReadCloser{
+		rc: rc,
+	}
+}
+
+func (b *byteCountReadCloser) Read(p []byte) (int, error) {
+	n, err := b.rc.Read(p)
+	atomic.AddInt64(&b.read, int64(n))
+	return n, err
+}
+
+func (b *byteCountReadCloser) Close() error {
+	return b.rc.Close()
+}
+
+func (b *byteCountReadCloser) WriteTo(w io.Writer) (int64, error) {
+	n, err := io.Copy(w, b.rc)
+	atomic.AddInt64(&b.read, n)
+	return n, err
+}
+
+func (b *byteCountReadCloser) BytesRead() int64 {
+	return b.read
```

Comment:
You should probably also here use atomic.

BTW, I [made it this way in my own code](https://gitlab.com/tozd/waf/-/blob/main/counters.go?ref_type=heads#L219-288).